pySDC/helpers/fieldsIO.py

```diff
@@ -624,9 +651,11 @@ def addField(self, time, field):
             self.MPI_WRITE(np.array(time, dtype=T_DTYPE))
         offset0 += self.tSize
 
+        _num_writes = 0
         for (iVar, *iBeg) in itertools.product(range(self.nVar), *[range(n) for n in self.nLoc[:-1]]):
```

Comment:
That's not exactly what I suggested ... all IO operation must be collective with _all ! It's just that we have to ensure that all processes are calling exactly the same number of time MPI_WRITE_AT or MPI_READ_AT. It means that you have to add after the first for loop :

```python
for _ in range(self.num_collective_IO - _num_writes):
    self.MPI_WRITE_AT(0, field[:0])
```
That should avoid the deadlock

Reply:
Why is that better? I expect little performance hit by non-collective IO here because there are few non-collective operations relative to total operations.

Reply:
Still counts ... and you can have a majority of non-collective operation with some decomposition. For instance, a 2D problem of size [10,32] decomposed in the first direction in 6 sub-domains => 4 processes have 2 points, 2 have 1 point. Because you need to write contiguous chunk of data in the file, for the first loop all are doing collective write for the first points, but for the second points 4 processes are doing a non-collective operation.

It's just better to always have all process doing collective write, with those who don't write anything doing a no-op, and the MPI implementation is hopefully well done enough to avoid losing anything on that.

Reply:
Actually, I found some inconsistent behaviour when reading and writing nothing on some tasks on my laptop. Namely, if I print something, it goes through, but if I don't, it deadlocks. I don't understand this behaviour. I suggest we leave this with some non-collective IO operations and if you want, you can fix it when you return from vacation.

Reply:
I found were the problem was : `self.num_collective_IO` must be called at the beginning of `readField` or `addField`. Because the first time the property is evaluated, it calls a `allreduce` but then the process having the maximum number of collective IO are still launching a `write_at_all` or a `read_at_all`, which creates the deadlock.

Now, the `allreduce` is done before any other collective operation, the tests did pass on my laptop. Let's see what happen with the github tests ...